packages/core/src/input/InputManager.ts

```diff
@@ -0,0 +1,320 @@
+    canvas.addEventListener("pointerup", this._pushEventList);
+    canvas.addEventListener("pointermove", this._pushEventList);
+    canvas.addEventListener("pointercancel", this._pushEventList);
+    canvas.addEventListener("pointerout", this._pushEventList);
+  }
+
+  /**
+   * Called when the engine is destroyed.
+   */
+  destroy(): void {
+    const { _canvas: canvas } = this;
+    canvas.removeEventListener("pointerdown", this._pushEventList);
+    canvas.removeEventListener("pointerup", this._pushEventList);
+    canvas.removeEventListener("pointermove", this._pushEventList);
+    canvas.removeEventListener("pointercancel", this._pushEventList);
+    canvas.removeEventListener("pointerout", this._pushEventList);
+    this._eventList.length = 0;
+    this._pointerList.length = 0;
+    this._pointers.length = 0;
+    this._input = null;
+    this._pointerIdToIndexMap = null;
+  }
+
+  /**
+   * Update pointer event, will be executed every frame.
+   * @internal
+   */
+  _update(): void {
+    /** Simulate a combined pointer. */
+    const { _eventList } = this;
+    const eventLen = _eventList.length;
+    /** Expressed in binary. */
+    let evtTypeList = 1;
+    if (eventLen > 0) {
+      for (let l = 0; l < eventLen; l++) {
+        const evt = _eventList[l];
+        switch (evt.type) {
+          case "pointerdown":
+            this._changePointer(PointerOperationType.Update, evt.pointerId, evt.pageX, evt.pageY);
+            /** this._activePointerCount === 1 && (evtTypeList = (evtTypeList << 1) | PointerEventType.Down); */
+            this._activePointerCount === 1 && (evtTypeList <<= 1);
+            break;
+          case "pointerup":
+            this._activePointerCount === 1 && (evtTypeList = (evtTypeList << 1) | PointerEventType.Up);
+            break;
+          case "pointermove":
+            this._changePointer(PointerOperationType.Update, evt.pointerId, evt.pageX, evt.pageY);
+            break;
+          case "pointercancel":
+          case "pointerout":
+            this._changePointer(PointerOperationType.Remove, evt.pointerId);
+            break;
+          default:
+            break;
+        }
+      }
+      /** Reset event list. */
+      _eventList.length = 0;
+    }
+
+    /** Sync _pointers and _pointerList. */
+    const { _pointers: pointers, _pointerList: pointerList, _input: input, _activePointerCount } = this;
+    pointers.length = _activePointerCount;
+    if (evtTypeList <= 1 && _activePointerCount === 0) {
+      return;
+    } else {
+      for (let i = 0; i < _activePointerCount; i++) {
+        pointers[i] = pointerList[i];
+      }
+    }
+
+    /** Check whether pressed events are triggered. */
+    let prePressedEntity = input.pressedEntity;
+    if (prePressedEntity) {
+      const preScripts = prePressedEntity._scripts._elements;
+      for (let i = preScripts.length - 1; i >= 0; i--) {
+        preScripts[i].onPointerDrag();
```

Comment:
`onPointerDrag` is fire when  hold down in before frame and still  holding down  this frame. So fire this event after raycast is Correctly.